operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1673 @@
+package lbipam
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/ipam/service/ipallocator"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+	core_v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	client_typed_v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	operatorOption "github.com/cilium/cilium/operator/option"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/k8s"
+	cilium_api_v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	cilium_client_v2alpha1 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_labels "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
+	slim_meta_v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/stream"
+)
+
+const (
+	// The condition added to services to indicate if a request for IPs could be satisfied or not
+	ciliumSvcRequestSatisfiedCondition = "io.cilium/lb-ipam-request-satisfied"
+
+	// The annotation LB IPAM will look for when searching for requested IPs
+	ciliumSvcLBIPSAnnotation = "io.cilium/lb-ipam-ips"
+
+	// The string used in the FieldManager field on update options
+	ciliumFieldManager = "cilium-operator-lb-ipam"
+
+	serviceNamespaceLabel = "io.kubernetes.service.namespace"
+	serviceNameLabel      = "io.kubernetes.service.name"
+)
+
+var Cell = cell.Module(
+	"LB-IPAM",
+	cell.Config(defaultLBIPAMConfig),
+	// Provide LBIPAM so instances of it can be used while testing
+	cell.Provide(registerLBIPAM),
+	// Invoke an empty function which takes an LBIPAM to force its construction.
+	cell.Invoke(func(*LBIPAM) {}),
```

Comment:
I'm still not super happy that for "standalone" modules we need this invoke. For these we kind of want the object construction gated by the configuration, e.g. we want to only invoke if `Enable == true`. This is not hard to implement: `cell.Instantiate[*LBIPAM](LBIPAMConfig.IsEnabled)`, internally would do an invoke first to get the answer to `IsEnabled` and then do the actual invoke to construct the object.

OTOH, for standalone things which don't expose a public API we could just have the invoke (no provide) that doesn't register any lifecycle hooks if disabled... but then we wouldn't have the object in the dependency graph so we wouldn't see its dependencies, so it's kind of nice to still have these "terminal" objects around.

Not sure what's preferred.

Reply:
I originally had just an Invoke, but I need access to the LBIPAM in the integration tests to inspect some internal state, which is why I went for this.

I am not sure about the implementation of `cell.Instantiate` since `LBIPAMConfig` is also an injected object. I currently do do the DI for LBIPAM but don't register a hook in the lifecycle if the component isn't enabled. This way the debug dot graph / objects is at least always consistent no matter the config.